consult-org.el

```diff
@@ -0,0 +1,167 @@
+         (setq org-outline-path-cache nil))
+       (setq line (+ line (consult--count-lines (prog1 (point)
+                                                  (goto-char opoint)))))
+       (setq opoint (point))
+       (pcase-let ((`(_ ,level ,todo ,prio) (org-heading-components)))
+         (consult--location-candidate
+          (org-format-outline-path (org-get-outline-path t t))
+          (point-marker)
+          line
+          'consult-org--level level
+          'consult-org--todo todo
+          'consult-org--priority prio)))
+     match scope skip)))
+
+;;;###autoload
+(defun consult-org-heading (&optional match scope)
+  "Jump to an Org heading.
+
+MATCH and SCOPE are as in `org-map-entries' and determine which
+entries are offered.  By default, all entries of the current
+buffer are offered."
+  (interactive (unless (derived-mode-p 'org-mode)
+                 (user-error "Must be called from an Org buffer.")))
+  (consult--jump
+   (consult--read
+    (consult--with-increased-gc (consult-org--entries match scope))
+    :prompt "Go to heading: "
+    :category 'consult-location
+    :sort nil
+    :title (unless (member scope '(nil 'tree 'region 'region-start-level 'file))
+             ;; Don't add titles when only showing entries from current buffer
+             (lambda (cand)
+               (let ((marker (car (get-text-property 0 'consult-location cand))))
+                 (buffer-name (marker-buffer marker)))))
+    :narrow (consult-org--narrow)
+    :require-match t
+    :lookup #'consult--lookup-location
+    :history '(:input consult-org--history)
+    :state (consult--jump-preview))))
+
+;;;###autoload
+(defun consult-org-agenda (&optional match)
+  "Jump to an Org agenda heading.
+
+By default, all agenda entries are offered. MATCH is as in
+`org-map-entries' and can used to refine this."
+  (interactive)
+  (unless org-agenda-files
+    (user-error "No agenda files"))
+  (consult-org-heading match 'agenda))
+
+;;;###autoload
+(defun consult-org-clock-in (&optional match scope)
+  "Clock into an Org heading.
+
+MATCH and SCOPE are as in `org-map-entries' and determine which
+entries are offered. By default, offer entries of files with a
+recent clocked item."
+  (interactive)
+  (setq scope (or scope
+                  (thread-last (progn (org-clock-load) org-clock-history)
+                    (mapcar 'marker-buffer)
+                    (mapcar 'buffer-file-name)
+                    seq-uniq
+                    (seq-remove 'null))
+                  (user-error "No recent clocked tasks")))
+  (org-clock-clock-in
+   (list
+    (consult--read
+     (consult--with-increased-gc
+      (sort ;; Recent items first, everything else in their original order
+       (consult-org--entries match scope)
+       (lambda (c1 c2)
+         (let ((m1 (car (get-text-property 0 'consult-location c1)))
+               (m2 (car (get-text-property 0 'consult-location c2))))
+           (and (member m1 org-clock-history)
```

Comment:
Same here, no member in the sorting because of `O(n*log(n)*m)`.

Reply:
But maybe the clock history is always very short or you could cut it off? As I said, I am not using the clocking feature yet. Generally I like to keep long histories.

Reply:
Okay, let's think about this. Sorting usually takes O(n log(n)) operations. But now each operation takes O(m), so that makes it O(mn log(n)).

An alterative would be to split off the recent items (takes O(mn)) and then sort only the recent items (takes O(m^2 log(m^2)), and concatenate (takes O(n)). Overall this is O(mn), assuming n >> m.

Replacing the list by a hash table would cut it down to O(n log(m)), but is probably doesn't break even in practice.

Should we go for the O(mn) option?

Reply:
The proper solution is to replace the list by the hash table to ensure O(nlogn), but if the list is guaranteed to be short ("constant") then we already have that. 

I am a bit wary about the quadratic behavior, since in my Vertico package sorting by history position turned out to be a big bottleneck, which had to be fixed using a hashtable. But there one does not control the history since the UI should work for arbitrary commands. Here we know the history.

You are right about the break even, the hashtable is slow and one should decorate the items before sorting to cut down factors a bit. But breaking even is not what matters, since for small numbers we are already fast in any case. Even when the optimization hurts, we are not hurt much. I am only asking about retaining good behavior for long histories.

Btw, there is already the helper `consult--string-hash` if one would decide to turn the list into a hash.

Reply:
Hmm, assuming you have these 20 element histories, member is good enough and we can always use `consult--string-hash` later on if it turns out to be a problem. 

But then - it seems this clocking history is not really a history but you intend to use it here as a list of recent items which are still relevant for clocking, the elements which got pushed over the edge are not relevant anymore. I must say I am not really fond of this design. I think I would prefer to have a more deterministically behaving command `consult-clock-in-agenda` which uses the agenda, which is controlled by the user. And additionally one can still sort using the history. The history can be long in that case, requiring the hashtable.

The problem here is that I am not using the clocking currently, otherwise I could judge a bit better.

Reply:
The basic way to clock in is to move to a headline and call `M-x org-clock-in`. The alternative is to type `C-u M-x org-clock-in`; then a window with 20 recent tasks appears and you can press a character to select one of them.

Originally I just wanted a command to clock into agenda entries. But now I realize that, if I were to clock into a task in a random file, then it would be quite likely that I would clock into another entry of that same file later on. On the other hand, I have 2 agenda files, and I only clock into items of one (the work related).

But maybe we should check what other people think. Perhaps @oantolin and @protesilaos have an opinion?